feat: compile sierra to casm in declare tx

ArniStarkware opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/starkware-libs/mempool/234)
<!-- Reviewable:end -->

ArniStarkware commented:
* **#235** <a href="https://app.graphite.dev/github/pr/starkware-libs/mempool/235?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#234** <a href="https://app.graphite.dev/github/pr/starkware-libs/mempool/234?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ArniStarkware and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Yael-Starkware reviewed (changes requested):
> Reviewed 3 of 6 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: 5 of 6 files reviewed, 13 unresolved discussions (waiting on @ArniStarkware and @dafnamatsry)

___
*[`crates/gateway/src/gateway.rs` line 129 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Fr9it8LFfMHTPfm9E:-O-Fr9it8LFfMHTPfm9F:undefined) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L129)):*
> ```XML
>     let optional_class_info = get_optional_class_info(&tx)?;
> 
>     // TODO(Yael, 19/5/2024): pass the relevant class_info and deploy_account_hash.
> ```





_Suggestion:_
```XML
pass the relevant deploy_account_hash.
```
___
*[`crates/gateway/src/gateway.rs` line 141 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Fs8Jf2HsVNUBKwqBY:-O-Fs8Jf2HsVNUBKwqBZ:undefined) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L141)):*
> ```XML
> 
> fn get_optional_class_info(tx: &ExternalTransaction) -> GatewayResult<Option<ClassInfo>> {
>     if let ExternalTransaction::Declare(ExternalDeclareTransaction::V3(tx)) = tx {
> ```





_Suggestion:_
```XML
let tx = match tx {
    ExternalTransaction::Declare(ExternalDeclareTransaction::V3(tx)) => tx,
    _ => return Ok(None)
}
```
___
*[`crates/gateway/src/gateway.rs` line 142 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-FtF6F3EGgp3BCl0Re:-O-FtF6G-OYjRfa9mu73:b6y5ox4) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L142)):*
> ```XML
> fn get_optional_class_info(tx: &ExternalTransaction) -> GatewayResult<Option<ClassInfo>> {
>     if let ExternalTransaction::Declare(ExternalDeclareTransaction::V3(tx)) = tx {
>         let starknet_api_contract_class: starknet_api::external_transaction::ContractClass =
> ```

Move to use statement 
you can set it as StarkNetApiContractClass, BlockifierContractCalss and CairoContractClass



_Code quote:_
```XML
starknet_api::external_transaction::ContractClass
```
___
*[`crates/gateway/src/gateway.rs` line 162 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Fu6uo1Q787wK5mEHA:-O-Fu6uo1Q787wK5mEHB:b-w4hy0t) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L162)):*
> ```XML
>         };
> 
>         // TODO: Handle unwrap.
> ```

Is this something you meant to do in this PR? 



_Code quote:_
```XML
// TODO: Handle unwrap.
```
___
*[`crates/gateway/src/gateway.rs` line 168 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-FuF0G6bBNhHCc3Cm9:-O-FuF0G6bBNhHCc3CmA:bqaxfi) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L168)):*
> ```XML
>             ContractClassV1::try_from_json_string(&raw_contract_class).unwrap();
> 
>         let blockifier_contract_class: blockifier::execution::contract_class::ContractClass =
> ```

move to use statement at the top



_Code quote:_
```XML
blockifier::execution::contract_class::ContractClass
```
___
*[`crates/gateway/src/gateway.rs` line 172 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G3M-78pXRz146UPEb:-O-G3M-78pXRz146UPEc:buugp6x) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L172)):*
> ```XML
>         let class_info =
>             ClassInfo::new(&blockifier_contract_class, sierra_program_length, abi_length)
>                 .expect("Expects a Cairo 1 contract class");
> ```

is that the only possible error?



_Code quote:_
```XML
.expect("Expects a Cairo 1 contract class");
```
___
*[`crates/gateway/src/gateway.rs` line 180 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G3nS_CWrkytkv_37x:-O-G3nS_CWrkytkv_37y:bb7od0r) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L180)):*
> ```XML
> 
> fn starknet_api_contract_class_to_cairo_lang_contract_class(
>     starknet_api_contract_class: starknet_api::external_transaction::ContractClass,
> ```

same



_Code quote:_
```XML
 starknet_api_contract_class: starknet_api::external_transaction::ContractClass
```
___
*[`crates/gateway/src/gateway.rs` line 181 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G3xlt7M-W3uylXuyX:-O-G3xlt7M-W3uylXuyY:bb7od0r) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L181)):*
> ```XML
> fn starknet_api_contract_class_to_cairo_lang_contract_class(
>     starknet_api_contract_class: starknet_api::external_transaction::ContractClass,
> ) -> cairo_lang_starknet_classes::contract_class::ContractClass {
> ```

same



_Code quote:_
```XML
 cairo_lang_starknet_classes::contract_class::ContractClass
```
___
*[`crates/gateway/src/gateway.rs` line 192 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G4PH2C6clopc-CCVe:-O-G4PH2C6clopc-CCVf:bb7od0r) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L192)):*
> ```XML
>     );
> 
>     cairo_lang_starknet_classes::contract_class::ContractClass {
> ```

same



_Code quote:_
```XML
cairo_lang_starknet_classes::contract_class::ContractClass
```
___
*[`crates/gateway/src/gateway.rs` line 198 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G4eXY7136aXjMFtm0:-O-G4eXY7136aXjMFtm1:b-imn8nw) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L198)):*
> ```XML
>         entry_points_by_type,
>         // The Abi is irrelevant to the computlation.
>         abi: None,
> ```

This could be error prone, if this function is used for a different purpose.
If complicated to implement I think it should be at least documented at the function level



_Code quote:_
```XML
        // The Abi is irrelevant to the computlation.
        abi: None,
```
___
*[`crates/gateway/src/gateway.rs` line 204 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G51TyAX5izY-YlSw6:-O-G51TyAX5izY-YlSw7:b-7jl281) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L204)):*
> ```XML
> fn entry_point_by_type_to_contract_entry_points(
>     entry_points_by_type: starknet_api::external_transaction::EntryPointByType,
> ) -> cairo_lang_starknet_classes::contract_class::ContractEntryPoints {
> ```

same



_Code quote:_
```XML
    entry_points_by_type: starknet_api::external_transaction::EntryPointByType,
) -> cairo_lang_starknet_classes::contract_class::ContractEntryPoints {
```
___
*[`crates/gateway/src/gateway.rs` line 234 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G7H9f-n5vkeTTcW40:-O-G7H9f-n5vkeTTcW41:b873y6m) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L234)):*
> ```XML
> 
> fn stark_felt_to_big_uint(stark_felt: StarkFelt) -> BigUint {
>     // TODO: Solve the unwrap.
> ```

I think you need to propagate the error. 



_Code quote:_
```XML
// TODO: Solve the unwrap.
```
___
*[`crates/starknet_sierra_compile/src/compile.rs` line 23 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-FrNuzDewFpUzbncb4:-O-FrNuzDewFpUzbncb5:b45h) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/starknet_sierra_compile/src/compile.rs#L23)):*
> ```XML
>     #[error(transparent)]
>     StarknetSierraCompilationError(#[from] StarknetSierraCompilationError),
>     #[error("Compilation have paniced")]
> ```





_Suggestion:_
```XML
Compilation panicked
```


<!-- Sent from Reviewable.io -->

Yael-Starkware reviewed (commented):
> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: all files reviewed, 13 unresolved discussions (waiting on @ArniStarkware and @dafnamatsry)



<!-- Sent from Reviewable.io -->

ArniStarkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: 4 of 6 files reviewed, 13 unresolved discussions (waiting on @dafnamatsry and @Yael-Starkware)

___
*[`crates/gateway/src/gateway.rs` line 129 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Fr9it8LFfMHTPfm9E:-O-VEy66-q8rGy7Ao-wi:bmfxgaw) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L129)):*
> ```XML
>     let optional_class_info = get_optional_class_info(&tx)?;
> 
>     // TODO(Yael, 19/5/2024): pass the relevant class_info and deploy_account_hash.
> ```

Done :)
___
*[`crates/gateway/src/gateway.rs` line 141 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Fs8Jf2HsVNUBKwqBY:-O-VFi383YxYdNkoMh6Y:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L141)):*
> ```XML
> 
> fn get_optional_class_info(tx: &ExternalTransaction) -> GatewayResult<Option<ClassInfo>> {
>     if let ExternalTransaction::Declare(ExternalDeclareTransaction::V3(tx)) = tx {
> ```

Done.
___
*[`crates/gateway/src/gateway.rs` line 142 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-FtF6F3EGgp3BCl0Re:-O-V_2nn-9lNf2R7QIw1:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L142)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

Move to use statement 
you can set it as StarkNetApiContractClass, BlockifierContractCalss and CairoContractClass
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 168 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-FuF0G6bBNhHCc3Cm9:-O-VarUm1rknyMZyaZQN:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L168)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

move to use statement at the top
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 180 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G3nS_CWrkytkv_37x:-O-V_WlL2hbpuFiKWABE:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L180)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 181 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G3xlt7M-W3uylXuyX:-O-V_X6j0VXn4I_OnZAP:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L181)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 192 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G4PH2C6clopc-CCVe:-O-VaW_49Fv7MNzRfmUU:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L192)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 204 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G51TyAX5izY-YlSw6:-O-VapxhBkYv2TOr1sGk:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L204)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 234 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G7H9f-n5vkeTTcW40:-O-VnmQ8CptVg0bW68Cz:bfwi0mg) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L234)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

I think you need to propagate the error.
</blockquote></details>

I looked into the function `BigUint::from_radix_be`.
It returns an `Optional(BigUint)`. The only case when it returns `None` is when `radix` (the second parameter) is *not* `256`.
___
*[`crates/starknet_sierra_compile/src/compile.rs` line 23 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-FrNuzDewFpUzbncb4:-O-Vamjl3zkTdaqcvNAl:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/starknet_sierra_compile/src/compile.rs#L23)):*
> ```XML
>     #[error(transparent)]
>     StarknetSierraCompilationError(#[from] StarknetSierraCompilationError),
>     #[error("Compilation have paniced")]
> ```

Done.


<!-- Sent from Reviewable.io -->

dafnamatsry reviewed (changes requested):
> Reviewed 1 of 6 files at r1, 1 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: 4 of 6 files reviewed, 24 unresolved discussions (waiting on @ArniStarkware and @Yael-Starkware)

___
*[`crates/gateway/src/errors.rs` line 16 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xpm4_10WzXeP0tgTc:-O-Xpm4_10WzXeP0tgTd:undefined) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/errors.rs#L16)):*
> ```XML
> pub enum GatewayError {
>     #[error(transparent)]
>     CompilationUtilError(#[from] starknet_sierra_compile::compile::CompilationUtilError),
> ```





_Suggestion:_
```XML
CompilationError
```
___
*[`crates/gateway/src/gateway.rs` line 198 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G4eXY7136aXjMFtm0:-O-XiTh01pC-OqzZAZAo:bms4d5j) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L198)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

This could be error prone, if this function is used for a different purpose.
If complicated to implement I think it should be at least documented at the function level
</blockquote></details>

Agreed. Consider renaming the function to `get_contract_class_for_compilation`.
___
*[`crates/gateway/src/gateway.rs` line 234 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G7H9f-n5vkeTTcW40:-O-XnyL09-Pcf78a6FWQ:b-4s4ob0) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L234)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

I looked into the function `BigUint::from_radix_be`.
It returns an `Optional(BigUint)`. The only case when it returns `None` is when `radix` (the second parameter) is *not* `256`.
</blockquote></details>

I think you have `BigUint::from_bytes_be()`.
___
*[`crates/gateway/src/gateway.rs` line 148 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-XeBjhBhY2-RPd3Hh0:-O-XeBji1r-_htv1mnqw:bdzslz7) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L148)):*
> ```XML
> }
> 
> fn get_optional_class_info(tx: &ExternalTransaction) -> GatewayResult<Option<ClassInfo>> {
> ```

Few things about the function signature:
* It should be clear from the name of the function that it compiles the contract class. `get_optional_class_info` doesn't describe anything about what the function does.
* No reason to get an `ExternalTransaction` if the function is only relevant for Declare transactions. And the optional in the return value is also not relevant anymore.
* Note that by changing the type of the argument, passing wrong tx type will be caught at compilation rather than by returning None at runtime.



_Suggestion:_
```XML
fn compile_contract_class(tx: &ExternalDeclareTransaction) -> GatewayResult<ClassInfo>
```
___
*[`crates/gateway/src/gateway.rs` line 182 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xp5OJ0K9ULB85vzBr:-O-Xp5OJ0K9ULB85vzBs:b-ouy3p5) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L182)):*
> ```XML
>     Ok(Some(class_info))
> }
> ```

Please move all the conversion code to some utils file. Maybe even to the compilation crate.
___
*[`crates/gateway/src/gateway.rs` line 183 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-XgvEQ7D57gNfFpWD3:-O-XgvER1Abb3fOeI_52:b-qygc25) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L183)):*
> ```XML
> }
> 
> fn starknet_api_contract_class_to_cairo_lang_contract_class(
> ```

It can be inferred from the argument that it is SNAPI contract class that is being converted.



_Suggestion:_
```XML
cairo_lang_contract_class
```
___
*[`crates/gateway/src/gateway.rs` line 184 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xjbh79f6FsAVuouBF:-O-Xjbh79f6FsAVuouBG:undefined) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L184)):*
> ```XML
> 
> fn starknet_api_contract_class_to_cairo_lang_contract_class(
>     starknet_api_contract_class: StarknetApiContractClass,
> ```





_Suggestion:_
```XML
starknet_api_contract_class: &StarknetApiContractClass
```
___
*[`crates/gateway/src/gateway.rs` line 191 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-XhmV3DVEeI6jxSTge:-O-XhmV3DVEeI6jxSTgf:b-492z1a) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L191)):*
> ```XML
>         .map(stark_felt_to_big_uint_as_hex)
>         .collect();
>     let contract_class_version = starknet_api_contract_class.contract_class_version;
> ```

You only use this variable once, so no need to define it here and better to just call `starknet_api_contract_class.contract_class_version` when needed.



_Code quote:_
```XML
let contract_class_version = starknet_api_contract_class.contract_class_version;
```
___
*[`crates/gateway/src/gateway.rs` line 201 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-XiNhuAvs_JjBxdUif:-O-XiNhuAvs_JjBxdUig:undefined) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L201)):*
> ```XML
>         contract_class_version,
>         entry_points_by_type,
>         // The Abi is irrelevant to the computlation.
> ```





_Suggestion:_
```XML
compilation
```
___
*[`crates/gateway/src/gateway.rs` line 206 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xjn6E5o_sySKlpJ-H:-O-Xjn6E5o_sySKlpJ-I:b45h) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L206)):*
> ```XML
> }
> 
> fn entry_point_by_type_to_contract_entry_points(
> ```





_Suggestion:_
```XML
cairo_lang_contract_entry_points
```
___
*[`crates/gateway/src/gateway.rs` line 207 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xk8EDDT25pau_SUjw:-O-Xk8EDDT25pau_SUjx:undefined) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L207)):*
> ```XML
> 
> fn entry_point_by_type_to_contract_entry_points(
>     entry_points_by_type: StarknetApiEntryPointByType,
> ```





_Suggestion:_
```XML
 entry_points_by_type: &StarknetApiEntryPointByType,
```
___
*[`crates/gateway/src/gateway.rs` line 219 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xleeg8uzu3nc9tQ-U:-O-Xleeg8uzu3nc9tQ-V:undefined) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L219)):*
> ```XML
> fn starknet_api_entry_points_to_contract_entry_points(
>     entry_points: Vec<StarknetApiEntryPoint>,
> ) -> Vec<CairoLangContractEntryPoint> {
> ```





_Suggestion:_
```XML
fn cairo_lang_contract_entry_points_vec(
    entry_points: &Vec<StarknetApiEntryPoint>,
) -> Vec<CairoLangContractEntryPoint> {
```
___
*[`crates/gateway/src/gateway.rs` line 225 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-XkrmP6iNIfXHkj6-3:-O-XkrmP6iNIfXHkj6-4:undefined) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L225)):*
> ```XML
> fn entry_point_into_contract_entry_point(
>     entry_point: StarknetApiEntryPoint,
> ) -> CairoLangContractEntryPoint {
> ```





_Suggestion:_
```XML
fn cairo_lang_contract_entry_point(
    entry_point: &StarknetApiEntryPoint,
) -> CairoLangContractEntryPoint {
```


<!-- Sent from Reviewable.io -->

ArniStarkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: 0 of 9 files reviewed, 16 unresolved discussions (waiting on @ayeletstarkware, @dafnamatsry, and @Yael-Starkware)

___
*[`crates/gateway/src/errors.rs` line 16 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xpm4_10WzXeP0tgTc:-O-_o3S9EKkr6n0hxvnL:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/errors.rs#L16)):*
> ```XML
> pub enum GatewayError {
>     #[error(transparent)]
>     CompilationUtilError(#[from] starknet_sierra_compile::compile::CompilationUtilError),
> ```

Done.
___
*[`crates/gateway/src/gateway.rs` line 162 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Fu6uo1Q787wK5mEHA:-O-aE5qXCWrYwQ_Vz_O1:b-p8zgog) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L162)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

Is this something you meant to do in this PR?
</blockquote></details>

Done. See similar precedent in #225. 
___
*[`crates/gateway/src/gateway.rs` line 172 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G3M-78pXRz146UPEb:-O-a8mzqAKRN8uUpwOuL:b-453bkf) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L172)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

is that the only possible error?
</blockquote></details>

Yes. Added explicit error handling: 
1. for readability
2. for the Rust compiler to tell us if new error variants are supported by this method.

@ayeletstarkware I see you were the last one to handle the `ClassInfo::new` method. Must it be this unforgiving?
___
*[`crates/gateway/src/gateway.rs` line 198 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G4eXY7136aXjMFtm0:-O-_cIxr5lE2GpgFVzOn:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L198)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Agreed. Consider renaming the function to `get_contract_class_for_compilation`.
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 234 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G7H9f-n5vkeTTcW40:-O-_dogj6Gynq5i4tDFh:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L234)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

I think you have `BigUint::from_bytes_be()`.
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 148 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-XeBjhBhY2-RPd3Hh0:-O-_O9c45vnXqFetmkie:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L148)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Few things about the function signature:
* It should be clear from the name of the function that it compiles the contract class. `get_optional_class_info` doesn't describe anything about what the function does.
* No reason to get an `ExternalTransaction` if the function is only relevant for Declare transactions. And the optional in the return value is also not relevant anymore.
* Note that by changing the type of the argument, passing wrong tx type will be caught at compilation rather than by returning None at runtime.
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 182 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xp5OJ0K9ULB85vzBr:-O-_tD6t4mcOYQRW46S3:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L182)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Please move all the conversion code to some utils file. Maybe even to the compilation crate.
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 183 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-XgvEQ7D57gNfFpWD3:-O-_o69T-BPpkYdBsYE3:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L183)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

It can be inferred from the argument that it is SNAPI contract class that is being converted.
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 184 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xjbh79f6FsAVuouBF:-O-_aZ2L9tF8_QQ8G3_s:b-15wlep) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L184)):*
> ```XML
> 
> fn starknet_api_contract_class_to_cairo_lang_contract_class(
>     starknet_api_contract_class: StarknetApiContractClass,
> ```

Done.
This was important. It had sound ripple effects.
___
*[`crates/gateway/src/gateway.rs` line 191 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-XhmV3DVEeI6jxSTge:-O-_bvSX5EggtyphawZc:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L191)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

You only use this variable once, so no need to define it here and better to just call `starknet_api_contract_class.contract_class_version` when needed.
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 201 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-XiNhuAvs_JjBxdUif:-O-_bz-Z6MIfrg9spmV_:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L201)):*
> ```XML
>         contract_class_version,
>         entry_points_by_type,
>         // The Abi is irrelevant to the computlation.
> ```

Done.
___
*[`crates/gateway/src/gateway.rs` line 206 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xjn6E5o_sySKlpJ-H:-O-aDc6W-vobOQxQL3mR:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L206)):*
> ```XML
> }
> 
> fn entry_point_by_type_to_contract_entry_points(
> ```

Done.
___
*[`crates/gateway/src/gateway.rs` line 207 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xk8EDDT25pau_SUjw:-O-_cMOf52NmZtgRuWdQ:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L207)):*
> ```XML
> 
> fn entry_point_by_type_to_contract_entry_points(
>     entry_points_by_type: StarknetApiEntryPointByType,
> ```

Done.
___
*[`crates/gateway/src/gateway.rs` line 219 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xleeg8uzu3nc9tQ-U:-O-_cdxSCb0XhhJsiIQ_:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L219)):*
> ```XML
> fn starknet_api_entry_points_to_contract_entry_points(
>     entry_points: Vec<StarknetApiEntryPoint>,
> ) -> Vec<CairoLangContractEntryPoint> {
> ```

Done.
___
*[`crates/gateway/src/gateway.rs` line 225 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-XkrmP6iNIfXHkj6-3:-O-_o9Q57mBhbYz4N8ky:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L225)):*
> ```XML
> fn entry_point_into_contract_entry_point(
>     entry_point: StarknetApiEntryPoint,
> ) -> CairoLangContractEntryPoint {
> ```

Done.


<!-- Sent from Reviewable.io -->

dafnamatsry reviewed (changes requested):
> Reviewed 1 of 2 files at r3, 2 of 8 files at r4, 1 of 3 files at r5, 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: all files reviewed, 8 unresolved discussions (waiting on @ArniStarkware, @ayeletstarkware, and @Yael-Starkware)

___
*[`crates/gateway/src/errors.rs` line 23 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05SrsjBDiHA-4msu8g:-O05SrsjBDiHA-4msu8h:b6wnr9d) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/errors.rs#L23)):*
> ```XML
>     MessageSendError(String),
>     #[error(transparent)]
>     ProgramError(#[from] ProgramError),
> ```

Is this error needed?



_Code quote:_
```XML
    #[error(transparent)]
    ProgramError(#[from] ProgramError),
```
___
*[`crates/gateway/src/gateway.rs` line 184 at r3](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-Xjbh79f6FsAVuouBF:-O05QJ1--S1dxDnvZIXT:b32vhw) ([raw file](https://github.com/starkware-libs/mempool/blob/af430602bc3cad3363dcaeb831b00256dc9a4e1f/crates/gateway/src/gateway.rs#L184)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Done.
This was important. It had sound ripple effects.
</blockquote></details>

:)
___
*[`crates/gateway/src/gateway.rs` line 162 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05O7Xn036UizlLdAYt:-O05O7Xn036UizlLdAYu:bmaamhx) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/gateway.rs#L162)):*
> ```XML
> 
>     // Convert Casm contract class to Starknet contract class directly.
>     let blockifier_contract_class = ContractClassV1::try_from(casm_contract_class)?.into();
> ```

Needed?



_Code quote:_
```XML
.into()
```
___
*[`crates/gateway/src/gateway.rs` line 173 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05Rghe1ilq971L4yvG:-O05Rghe1ilq971L4yvH:b-udw0kn) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/gateway.rs#L173)):*
> ```XML
>             sierra_program_length: _,
>         }) => {
>             panic!(
> ```

Why panic?
The sierra program is an input from the user, and can be invalid and with 0 length.



_Code quote:_
```XML
panic!(
```


<!-- Sent from Reviewable.io -->

ArniStarkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: all files reviewed, 8 unresolved discussions (waiting on @ayeletstarkware, @dafnamatsry, and @Yael-Starkware)

___
*[`crates/gateway/src/errors.rs` line 23 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05SrsjBDiHA-4msu8g:-O07CkWo2azwP6foQVBe:b1liylp) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/errors.rs#L23)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Is this error needed?
</blockquote></details>

Yes. The following line needs it:
```
ContractClassV1::try_from(casm_contract_class)?
```

___
*[`crates/gateway/src/gateway.rs` line 162 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05O7Xn036UizlLdAYt:-O07CzWY3rV5lKPC3QaM:bppaamq) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/gateway.rs#L162)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Needed?
</blockquote></details>

Yes.
`blockifier_contract_class` needs to be of type `ContractClass` (`blockifier::execution::contract_class::ContractClass`). This is an enum where one of it's members is of type: `V1`, which holds a `ContractClassV1` (`blockifier::execution::contract_class::ContractClassV1`).
___
*[`crates/gateway/src/gateway.rs` line 173 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05Rghe1ilq971L4yvG:-O07E1xG0PxlvuywpaUt:b-wenib9) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/gateway.rs#L173)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Why panic?
The sierra program is an input from the user, and can be invalid and with 0 length.
</blockquote></details>

The function `ClassInfo::new` allows the creation of class info for Cairo 0 and Cairo 1 contracts. 
The only way `ClassInfo::new` can have an error is when there is a mismatch. In this context, we are dealing with the compilation of Cairo 1 contracts (we are compiling Sierra -> Casm). It is an assumption that this error can never happen. We could have even used an `expect` instead of a match, but I wanted to ensure we are explicit here. If `ClassInfo::new` will add a new error variant - the compiler will tell us about it.


<!-- Sent from Reviewable.io -->

dafnamatsry reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: all files reviewed, 8 unresolved discussions (waiting on @ArniStarkware, @ayeletstarkware, and @Yael-Starkware)

___
*[`crates/gateway/src/errors.rs` line 23 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05SrsjBDiHA-4msu8g:-O08-KPtAF_0kscQNZeM:b-6p82yy) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/errors.rs#L23)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Yes. The following line needs it:
```
ContractClassV1::try_from(casm_contract_class)?
```

</blockquote></details>

It's not clear in the context of the `GatewayError`, to what program we refer to here. Maybe rename to`ContractProgramError` or similar.
___
*[`crates/gateway/src/gateway.rs` line 162 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05O7Xn036UizlLdAYt:-O080XCR3vLGRLRFRjfw:b-rjfbig) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/gateway.rs#L162)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Yes.
`blockifier_contract_class` needs to be of type `ContractClass` (`blockifier::execution::contract_class::ContractClass`). This is an enum where one of it's members is of type: `V1`, which holds a `ContractClassV1` (`blockifier::execution::contract_class::ContractClassV1`).
</blockquote></details>

In that case, would `ContractClass::V1(ContractClassV1::try_from(casm_contract_class)?)` work?
It is more explicit and clear what type we expect
___
*[`crates/gateway/src/gateway.rs` line 173 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05Rghe1ilq971L4yvG:-O082dFrEmRAVEowvXa4:b-8qybyp) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/gateway.rs#L173)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

The function `ClassInfo::new` allows the creation of class info for Cairo 0 and Cairo 1 contracts. 
The only way `ClassInfo::new` can have an error is when there is a mismatch. In this context, we are dealing with the compilation of Cairo 1 contracts (we are compiling Sierra -> Casm). It is an assumption that this error can never happen. We could have even used an `expect` instead of a match, but I wanted to ensure we are explicit here. If `ClassInfo::new` will add a new error variant - the compiler will tell us about it.
</blockquote></details>

Not sure the assumption that this can never happen is right.
We simply pass the length of `sierra_program` which is given to us by the user as part of `tx`. It can be 0.
You can say that we assume that if the compilation passed by now then it must be > 0. But I won't count on it.

I think we should return the error here.


<!-- Sent from Reviewable.io -->

ArniStarkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: 5 of 9 files reviewed, 8 unresolved discussions (waiting on @ayeletstarkware, @dafnamatsry, and @Yael-Starkware)

___
*[`crates/gateway/src/errors.rs` line 23 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05SrsjBDiHA-4msu8g:-O08h6XbDz9_4U9tO56C:bboxofy) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/errors.rs#L23)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

It's not clear in the context of the `GatewayError`, to what program we refer to here. Maybe rename to`ContractProgramError` or similar.
</blockquote></details>

Done. Changed to `DeclaredContractProgramError`.
___
*[`crates/gateway/src/gateway.rs` line 162 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05O7Xn036UizlLdAYt:-O08hnAO2DPdtz08CP0g:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/gateway.rs#L162)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

In that case, would `ContractClass::V1(ContractClassV1::try_from(casm_contract_class)?)` work?
It is more explicit and clear what type we expect
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway.rs` line 173 at r6](https://reviewable.io/reviews/starkware-libs/mempool/234#-O05Rghe1ilq971L4yvG:-O09OkdnD3oGgnmZRr0z:b2tegr5) ([raw file](https://github.com/starkware-libs/mempool/blob/c92756b662463660e6d399474c0e51ed6d3e67ba/crates/gateway/src/gateway.rs#L173)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Not sure the assumption that this can never happen is right.
We simply pass the length of `sierra_program` which is given to us by the user as part of `tx`. It can be 0.
You can say that we assume that if the compilation passed by now then it must be > 0. But I won't count on it.

I think we should return the error here.
</blockquote></details>

Done. Added `DeclaredContractClassError`.


<!-- Sent from Reviewable.io -->

dafnamatsry reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234#-:-O0CsX4EAYG6qM7LgAKq:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @ayeletstarkware and @Yael-Starkware)



<!-- Sent from Reviewable.io -->

Yael-Starkware reviewed (approved):
> Reviewed 1 of 2 files at r3, 2 of 8 files at r4, 1 of 3 files at r5, 1 of 5 files at r6, 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: all files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware)



<!-- Sent from Reviewable.io -->

ArniStarkware commented:
*[`crates/gateway/src/gateway.rs` line 198 at r2](https://reviewable.io/reviews/starkware-libs/mempool/234#-O-G4eXY7136aXjMFtm0:-O0DZXHW7-OuqXg3Mrpg:bj6f8ks) ([raw file](https://github.com/starkware-libs/mempool/blob/7557759ba2ca1bc31d6a3caea5d3f6e6492b80ed/crates/gateway/src/gateway.rs#L198)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I renamed the function and added the required documentation.

Look for `into_contract_class_for_compilation` in `utils.rs` in this PR.


<!-- Sent from Reviewable.io -->

dafnamatsry reviewed (approved):
> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: all files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware and @Yael-Starkware)



<!-- Sent from Reviewable.io -->

Yael-Starkware reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/234)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @ayeletstarkware)



<!-- Sent from Reviewable.io -->
